### PR TITLE
Story/vspc 147

### PR DESCRIPTION
**Pull request by pradeep-017**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

Bibliography block on slide page: A user should be able to add new entries to the block (entering author, title, year, journal, etc information)
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/BiblioBlock.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+@Entity
+public class BiblioBlock extends ContentBlock implements IBiblioBlock {
+
+    private String title;
+    
+    private String author;
+    
+    private Integer year;
+    
+    private String journal;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/BiblioBlock.java:21]: for now, let's add at least the following:
- url 
- volume
- issue
- pages
- editors
- type
- note

we'll need to come up with a more flexible schema at some point, but let's leave it with this for now

File: vspace/src/main/webapp/WEB-INF/views/staff/modules/slides/slide.html [hunk 2]
```diff
@@ -65,7 +65,32 @@
...
+        $("#editAuthorBiblio").val(author);
+        $("#editYearBiblio").val(year);
+        $("#editJournalBiblio").val(journal);
+    	
+    	//remove card border
+        $(".open").css('display', 'none');
+        $(".open").css('overflow-y', 'scroll');
+        $(".open").css('max-height', '200px');
+    	
+    }
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/modules/slides/slide.html:92]: indentation for the above code

File: vspace/src/main/webapp/WEB-INF/views/staff/modules/slides/slide.html [hunk 3]
```diff
@@ -616,13 +776,62 @@
...
+           },
+           error: function(data) {
+        	   console.log(data)
+        		if (data.status == 403) {
+           			$("#loginAlert").show();
+           		}
+           		else {
+               		var alert = $('<div class="alert alert-danger alert-dismissible fade show" role="alert"><p>We are sorry but something went wrong. Please try again later.</p><button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button></div>');
+               		$('.error').append(alert);
+           		}
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/modules/slides/slide.html:823]: indentation

**Review by jdamerow — CHANGES_REQUESTED** [review]
test cases are missing (although you checked the box that you added them as appropriate)

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- the add bibliography block button should not have an image icon.
- the input fields need to be visible when entering blibographic information
- year should be a text field
- the input fields need to be longer
- the modal doesn't scroll, so if my window is too small I can't fill all fields (I actually can't click save)
- after adding a new block, the block is smaller than the rest

from the ticket:
"After a new bibliography block has been created, the user should be able to add new entries to the block (entering author, title, year, journal, etc information). The information would then be rendered according to a specific style (for now just one default one, but in the future we want to be able to offer other styles)."
so one block should have multiple citations.

also:
bibliography blocks are not shown on public page
"On the public page, Zotero's or Mendeley's browser plugin should recognize the entries and allow users to add them to their bibliography managers."

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteBiblioBlockController.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+@Controller
+public class DeleteBiblioBlockController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private IContentBlockManager contentBlockManager;
+
+    @RequestMapping(value = "/staff/module/{moduleId}/slide/{id}/biblio/{blockId}", method = RequestMethod.DELETE)
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteBiblioBlockController.java:26]: what if module with the id or slide with the id doesn't exist? currently that shouldn't be an issue, but that seems like a future security issue.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditBiblioBlockController.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import edu.asu.diging.vspace.core.model.impl.BiblioBlock;
+import edu.asu.diging.vspace.core.services.IContentBlockManager;
+
+@Controller
+public class EditBiblioBlockController {
+
+    @Autowired
+    private IContentBlockManager contentBlockManager;
+
+    @RequestMapping(value = "/staff/module/{moduleId}/slide/{id}/biblio/edit", method = RequestMethod.POST)
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditBiblioBlockController.java:24]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/BiblioBlock.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    @Transient
+    public String htmlRenderedBiblio() {
+        Parser parser = Parser.builder().build();
+        Node document = parser.parse(toString());
+        HtmlRenderer renderer = HtmlRenderer.builder().build();
+        return renderer.render(document);
+    }
+
+    @Transient
+    public String urlEncodedBiblioMetaData() throws UnsupportedEncodingException {
```
[the remark this answers is not in the document]
Comment (pradeep-017) [reply vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/BiblioBlock.java:168]: Can you suggest me how can we design this?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java [hunk 7]
```diff
@@ -266,4 +276,40 @@ public IChoiceBlock createChoiceBlock(String slideId, List<String> selectedChoic
         return choiceBlockRepo.save((ChoiceBlock)choiceBlock);
     }
 
+    @Override
+    public IBiblioBlock createBiblioBlock(String slideId, IBiblioBlock biblioData) {
+        ISlide slide = slideManager.getSlide(slideId);
+        IBiblioBlock biblioBlock = biblioBlockFactory.createBiblioBlock(slide, biblioData);
+        biblioBlock = biblioBlockRepo.save((BiblioBlock) biblioBlock);
+        return biblioBlock;
```
[the remark this answers is not in the document]
Comment (pradeep-017) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java:284]: Done!

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddBiblioBlockController.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Autowired
+    private IContentBlockManager contentBlockManager;
+
+    @RequestMapping(value = "/staff/module/{moduleId}/slide/{id}/bibliography", method = RequestMethod.POST)
+    public ResponseEntity<String> addTextBlock(@PathVariable("id") String slideId,
+            @PathVariable("moduleId") String moduleId, @RequestBody BiblioBlock biblioBlockData) throws IOException {
+
+        IBiblioBlock biblioBlock = contentBlockManager.createBiblioBlock(slideId, biblioBlockData);
+        return new ResponseEntity<>(biblioBlock.getId(), HttpStatus.OK);
```
[the remark this answers is not in the document]
Comment (pradeep-017) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddBiblioBlockController.java:29]: Done!

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteBiblioBlockController.java (hunk 4 above)
Comment (pradeep-017) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteBiblioBlockController.java:26]: I have taken of these checks.

**Comment by jdamerow** [comment]
as discussed, this needs to be changed.

**Pull request closed without merging**